include/swift/AST/ActorIsolation.h

```diff
@@ -191,7 +193,14 @@ class ActorIsolation {
 ActorIsolation getActorIsolation(ValueDecl *value);
 
 /// Determine how the given declaration context is isolated.
-ActorIsolation getActorIsolationOfContext(DeclContext *dc);
+/// \p getClosureActorIsolation allows the specification of actor isolation for
+/// closures that haven't been saved been saved to the AST yet. This is useful
+/// for solver-based code completion which doesn't modify the AST but stores the
+/// actor isolation of closures in the constraint system solution.
+ActorIsolation getActorIsolationOfContext(
+    DeclContext *dc,
+    llvm::function_ref<ClosureActorIsolation(AbstractClosureExpr *)>
+        getClosureActorIsolation);
```

Comment:
Should this be defaulted to `[](AbstractClosureExpr *closure) { return closure->getActorIsolation(); }` so 99% of the call sites don't have to repeat it?

Reply:
I wanted to do the same but that would mean that we need to include `Expr.h` here (because we are referencing into `AbstractClosureExpr`) and `Expr.h` transitively includes `ActorIsolation.h` creating an include dependency cycle. 😬

Reply:
Couldn't it default to the forward-declaration of `__AbstractClosureExpr_getActorIsolation`?

Reply:
Now that you mention it, I think that should be possible.

Reply:
🤞 Yeah, it should work with forward declation.

Reply:
Forward declaration also doesn’t work because `ClosureActorIsolation` is forward-declared here and the return type of `__AbstractClosureExpr_getActorIsolation`. I think that’s why clang complains
```
/Users/alex/swift-src/swift/include/swift/AST/ActorIsolation.h:207:9: error: no viable conversion from 'swift::ClosureActorIsolation (swift::AbstractClosureExpr *)' to 'llvm::function_ref<ClosureActorIsolation (AbstractClosureExpr *)>'
        getClosureActorIsolation = __AbstractClosureExpr_getActorIsolation);
        ^                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/alex/swift-src/llvm-project/llvm/include/llvm/ADT/STLExtras.h:171:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'swift::ClosureActorIsolation (swift::AbstractClosureExpr *)' to 'const llvm::function_ref<swift::ClosureActorIsolation (swift::AbstractClosureExpr *)> &' for 1st argument
class function_ref<Ret(Params...)> {
      ^
/Users/alex/swift-src/llvm-project/llvm/include/llvm/ADT/STLExtras.h:171:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'swift::ClosureActorIsolation (swift::AbstractClosureExpr *)' to 'llvm::function_ref<swift::ClosureActorIsolation (swift::AbstractClosureExpr *)> &&' for 1st argument
/Users/alex/swift-src/llvm-project/llvm/include/llvm/ADT/STLExtras.h:183:3: note: candidate constructor not viable: no known conversion from 'swift::ClosureActorIsolation (swift::AbstractClosureExpr *)' to 'std::nullptr_t' (aka 'nullptr_t') for 1st argument
  function_ref(std::nullptr_t) {}
  ^
/Users/alex/swift-src/llvm-project/llvm/include/llvm/ADT/STLExtras.h:186:3: note: candidate template ignored: requirement 'std::is_void<swift::ClosureActorIsolation>::value || std::is_convertible<swift::ClosureActorIsolation, swift::ClosureActorIsolation>::value' was not satisfied [with Callable = swift::ClosureActorIsolation (&)(swift::AbstractClosureExpr *)]
  function_ref(
  ^
/Users/alex/swift-src/swift/include/swift/AST/ActorIsolation.h:207:9: note: passing argument to parameter 'getClosureActorIsolation' here
        getClosureActorIsolation = __AbstractClosureExpr_getActorIsolation);
        ^

```

Reply:
I think it should be possible to move ClosureActorIsolation from Expr.h to ActorIsolation.h?...

Reply:
Not easily because it references `VarDecl *` inside a `PointerUnion`, which requires knowledge about the type of `VarDecl *` to compute `NumLowBitsAvailable` for it.

Reply:
Alright then :)

Reply:
Huh, TIL `function_ref` has SFINAE to allow covariant returns. Shame it doesn't have a constructor that just takes a callable with the exact params/return type. If we really wanted to, we could probably work around it by forward-declaring a trampoline that returns the `llvm::function_ref<ClosureActorIsolation(AbstractClosureExpr *)>`, but maybe it's not worth it.

Reply:
I think that’s a good idea 👍